cpp/src/arrow/buffer-builder.h

```diff
@@ -74,59 +74,63 @@ class ARROW_EXPORT BufferBuilder {
   /// \brief Ensure that builder can accommodate the additional number of bytes
   /// without the need to perform allocations
   ///
-  /// \param size number of additional bytes to make space for
+  /// \param additional_bytes number of additional bytes to make space for
   /// \return Status
-  Status Reserve(const int64_t size) { return Resize(size_ + size, false); }
+  Status Reserve(const int64_t additional_bytes) {
+    return Resize(size_ + additional_bytes, false);
+  }
+
+  /// \brief Return a capacity expanded by a growth factor of 2
+  static int64_t GrowByFactor(const int64_t min_capacity) {
+    // If the capacity was not already a multiple of 2, do so here
+    // TODO(emkornfield) doubling isn't great default allocation practice
+    // see https://github.com/facebook/folly/blob/master/folly/docs/FBVector.md
+    // for discussion
+    return BitUtil::NextPower2(min_capacity);
+  }
 
   /// \brief Append the given data to the buffer
   ///
   /// The buffer is automatically expanded if necessary.
-  Status Append(const void* data, int64_t length) {
-    if (capacity_ < length + size_) {
-      int64_t new_capacity = BitUtil::NextPower2(length + size_);
-      ARROW_RETURN_NOT_OK(Resize(new_capacity));
-    }
+  Status Append(const void* data, const int64_t length) {
+    ARROW_RETURN_NOT_OK(Resize(GrowByFactor(length + size_), false));
     UnsafeAppend(data, length);
     return Status::OK();
   }
 
+  /// \brief Append copies of a value to the buffer
+  ///
+  /// The buffer is automatically expanded if necessary.
+  Status Append(const int64_t num_copies, uint8_t value) {
+    ARROW_RETURN_NOT_OK(Resize(GrowByFactor(num_copies + size_), false));
```

Comment:
In `Resize`, `buffer_->Resize` is a virtual method that will be called everytime even if no resizing is necessary.